lib/modules/manager/sbt/extract.ts

```diff
@@ -335,11 +385,121 @@ export function extractPackageFile(
+  let scalaVersion: string | undefined = undefined;
+
+  for (const { packageFile, content } of packageFilesContent) {
+    const acc: PackageFile & ParseOptions = {
+      deps: [],
+      registryUrls,
+      localVars: globalVars,
+      globalVars: {},
+      packageFile,
+    };
+    const res = extractPackageFile(content, acc);
+
+    if (res) {
+      globalVars = { ...globalVars, ...res.localVars };
+      if (res.registryUrls) {
+        registryUrls = [...registryUrls, ...new Set(res.registryUrls)];
+      }
+      if (res.scalaVersion) {
+        scalaVersion = res.scalaVersion;
+      }
+    }
+  }
+
+  return {
+    globalVars,
+    registryUrls,
+    scalaVersion,
+  };
+}
+
+export async function extractAllPackageFiles(
+  _config: ExtractConfig,
+  packageFiles: string[]
+): Promise<PackageFile[] | null> {
+  // Read packages and store in groupPackageFileContent
+  // group package file by its folder
+  const groupPackageFileContent: GroupFilenameContent = {};
+  for (const packageFile of packageFiles) {
+    const content = await readLocalFile(packageFile, 'utf8');
+    if (!content) {
+      logger.trace({ packageFile }, 'packageFile has no content');
+      continue;
+    }
+    const group = upath.dirname(packageFile);
+    groupPackageFileContent[group] ??= [];
+    groupPackageFileContent[group].push({ packageFile, content });
+  }
+
+  // 1. globalVariables from project/ and root package file
+  // 2. registry from all package file
+  // 3. Project's scalaVersion - use in parseDepExpr to add suffix eg. "_2.13"
+  const { globalVars, registryUrls, scalaVersion } = prepareLoadPackageFiles([
+    ...(groupPackageFileContent['project'] ?? []), // in project/ folder
+    ...(groupPackageFileContent['.'] ?? []), // root
+  ]);
+
+  const mapDepsToPackageFile: Record<string, PackageDependency[]> = {};
+  // Start extract all package files
+  for (const packageFileContents of Object.values(groupPackageFileContent)) {
+    // Extract package file by its group
+    // local variable is share within its group
+    for (const { packageFile, content } of packageFileContents) {
+      const res = extractPackageFile(content, {
+        registryUrls,
+        deps: [],
+        packageFile,
+        scalaVersion,
+        localVars: {},
+        globalVars,
+      });
+      if (res) {
+        if (res?.deps) {
+          for (const dep of res.deps) {
+            const variableSourceFile = dep?.editFile ?? packageFile;
+            dep.registryUrls = [...new Set(dep.registryUrls)];
+            mapDepsToPackageFile[variableSourceFile] ??= [];
+            mapDepsToPackageFile[variableSourceFile].push(dep);
+          }
+        }
+      }
+    }
   }
 
-  return { deps, packageFileVersion };
+  // Filter unique package
+  // As we merge all package to single package file
+  // Packages are counted in submodule but it's the same one
+  // by packageName and currentValue
+  const finalPackages = Object.entries(mapDepsToPackageFile).map(
```

Comment:
This is a pain to debug, please rewrite as a regular loop, filling up a new array.

Also the nesting scares me off. Better aim for a duplicate check like this:
```ts
      const depAlreadyInPkgFile = pkgFile.deps.some((item) => ...);
      if (!depAlreadyInPkgFile) {
        pkgFile.deps.push(dep);
      }
```